tool/tctl/common/idp_command.go

```diff
@@ -0,0 +1,268 @@
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package common
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/alecthomas/kingpin/v2"
+	"github.com/gravitational/trace"
+	kyaml "k8s.io/apimachinery/pkg/util/yaml"
+
+	"github.com/gravitational/teleport"
+	samlidpv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/samlidp/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/asciitable"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/service/servicecfg"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// subcommandRunner is used to create pluggable subcommand under
+// idp command. E.g:
+// $ tctl idp saml <command> [<args> ...]
+// $ tctl idp oidc <command> [<args> ...]
+type subcommandRunner interface {
+	initialize(parent *kingpin.CmdClause, cfg *servicecfg.Config)
+	tryRun(ctx context.Context, selectedCommand string, c auth.ClientI) (match bool, err error)
+}
+
+// IdPCommand implements all commands under "tctl idp".
+type IdPCommand struct {
+	subcommandRunners []subcommandRunner
+}
+
+// samlIdPCommand implements all commands under "tctl idp saml"
+type samlIdPCommand struct {
+	cmd *kingpin.CmdClause
+
+	testAttributeMapping testAttributeMapping
+}
+
+// Initialize installs the base "idp" command and all subcommands.
+func (t *IdPCommand) Initialize(app *kingpin.Application, cfg *servicecfg.Config) {
+	idp := app.Command("idp", "Teleport Identity Provider")
+
+	idp.Alias(`
+Examples:
+  > tctl idp saml
+`)
+
+	t.subcommandRunners = []subcommandRunner{
+		&samlIdPCommand{},
+	}
+
+	for _, subcommandRunner := range t.subcommandRunners {
+		subcommandRunner.initialize(idp, cfg)
+	}
+}
+
+// TryRun calls tryRun for each subcommand, and returns (false, nil) if none of them match.
+func (i *IdPCommand) TryRun(ctx context.Context, cmd string, c auth.ClientI) (match bool, err error) {
+	for _, subcommandRunner := range i.subcommandRunners {
+		match, err = subcommandRunner.tryRun(ctx, cmd, c)
+		if err != nil {
+			return match, trace.Wrap(err)
+		}
+		if match {
+			return match, nil
+		}
+	}
+	return false, nil
+}
+
+func (s *samlIdPCommand) initialize(parent *kingpin.CmdClause, cfg *servicecfg.Config) {
+	samlcmd := parent.Command("saml", "SAML Identity Provider")
+	samlcmd.Alias(`
+Examples:
+
+  Test attribute mapping with given user and service provider.
+  > tctl idp saml test_attribute_mapping --users user.yaml (user spec file or username) --sp sp.yaml (service provider spec file) --format (json,yaml)
+`)
+	s.cmd = samlcmd
+
+	testAttrMap := samlcmd.Command("test_attribute_mapping", "Test expression evaluation of attribute mapping.")
```

Comment:
Nit: Could we make it `test-attribute-mapping` instead? I think we prefer to use dashes rather than underscores in CLI.

Reply:
renamed `test_attribute_mapping` to `test-attribute-mapping` here https://github.com/gravitational/teleport/pull/36360/commits/60771962126b20c3d948adac7461e9ed83902777